core/trino-main/src/main/java/io/trino/server/testing/exchange/LocalFileSystemExchangeSink.java

```diff
@@ -0,0 +1,156 @@
+
+    public static final String COMMITTED_MARKER_FILE_NAME = "committed";
+    public static final String DATA_FILE_SUFFIX = ".data";
+
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(LocalFileSystemExchangeSink.class).instanceSize();
+    private static final int INTEGER_INSTANCE_SIZE = ClassLayout.parseClass(Integer.class).instanceSize();
+
+    private static final int BUFFER_SIZE_IN_BYTES = toIntExact(DataSize.of(4, KILOBYTE).toBytes());
+
+    private final Path outputDirectory;
+    private final int outputPartitionCount;
+
+    @GuardedBy("this")
+    private final Map<Integer, SliceOutput> outputs = new HashMap<>();
+    @GuardedBy("this")
+    private boolean committed;
+    @GuardedBy("this")
+    private boolean closed;
+
+    public LocalFileSystemExchangeSink(Path outputDirectory, int outputPartitionCount)
+    {
+        this.outputDirectory = requireNonNull(outputDirectory, "outputDirectory is null");
+        this.outputPartitionCount = outputPartitionCount;
+    }
+
+    @Override
+    public synchronized void add(int partitionId, Slice data)
+    {
+        checkArgument(partitionId < outputPartitionCount, "partition id is expected to be less than %s: %s", outputPartitionCount, partitionId);
+        checkState(!committed, "already committed");
+        if (closed) {
+            return;
+        }
+        SliceOutput output = outputs.computeIfAbsent(partitionId, this::createOutput);
+        output.writeInt(data.length());
+        output.writeBytes(data);
+    }
+
+    private SliceOutput createOutput(int partitionId)
+    {
+        Path outputPath = outputDirectory.resolve(partitionId + DATA_FILE_SUFFIX);
+        try {
+            return new OutputStreamSliceOutput(new FileOutputStream(outputPath.toFile()), BUFFER_SIZE_IN_BYTES);
+        }
+        catch (FileNotFoundException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @Override
+    public synchronized long getSystemMemoryUsage()
+    {
+        return INSTANCE_SIZE +
+                estimatedSizeOf(outputs, (ignored) -> INTEGER_INSTANCE_SIZE, SliceOutput::getRetainedSize);
+    }
+
+    @Override
+    public synchronized void finish()
+    {
+        if (closed) {
+            return;
+        }
+        try {
+            for (SliceOutput output : outputs.values()) {
+                try {
+                    output.close();
+                }
+                catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            }
+            try {
+                createFile(outputDirectory.resolve(COMMITTED_MARKER_FILE_NAME));
+            }
+            catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+        catch (Throwable t) {
+            abort();
+            throw t;
+        }
+        committed = true;
+        closed = true;
```

Comment:
It may not matter in practice, as the `LocalFileSystemExchangeSink` object is likely to get discarded. Though I agree it is more correct to cleanup the data structures on finish. Added `outputs.clear()`.